### PR TITLE
[Merged by Bors] - Port bevy_ui to pipelined-rendering

### PR DESCRIPTION
**Pull request by Davier**
# Objective

Port bevy_ui to pipelined-rendering (see #2535 )

## Solution

I did some changes during the port:
- [X] separate color from the texture asset (as suggested [here](https://discord.com/channels/691052431525675048/743663924229963868/874353914525413406))
- [X] ~give the vertex shader a per-instance buffer instead of per-vertex buffer~ (incompatible with batching)

Remaining features to implement to reach parity with the old renderer:
- [x] textures
- [X] TextBundle

I'd also like to add these features, but they need some design discussion:
- [x] batching
- [ ] separate opaque and transparent phases
- [ ] multiple windows
- [ ] texture atlases
- [ ] (maybe) clipping

### CODE REVIEW

**Comment by cart** [comment]
Brilliant! I'll take a look at this ASAP.

In regard to your second and third lists:

Fortunately we already have a [draft MSAA implementation](#2541).

I'm cool with entirely ignoring the third list in the description. I'd prefer to not focus any time on adding new features to the current Bevy UI system, as thats next on our list of things to rework.

**Comment by alice-i-cecile** [comment]
For batching, perhaps coordinate with #2642?

**Comment by cart** [comment]
> For batching, perhaps coordinate with #2642?

I'd honestly rather just punt it unless @Davier is extremely motivated. We should invest the smallest amount of time possible on "legacy stuff".

**Comment by Davier** [comment]
> For batching, perhaps coordinate with #2642?

I've taken a lot of inspiration from that PR already, so implementing batching should be straightforward. I'm happy to do it (and any other point in the 3rd list) if you agree to review it, but I understand that you have a limited bandwidth.

> I'd honestly rather just punt it unless @Davier is extremely motivated. We should invest the smallest amount of time possible in "legacy stuff".

I'm doing it to learn so it's fine even if we throw it away soon. Or is there is another part of the renderer that I could help with?

File: examples/README.md
Comment (alice-i-cecile) [inline examples/README.md:251]:
```suggestion
`ui_pipelined` | [`ui/ui_pipelined.rs`](./ui/ui_pipelined.rs) | Illustrates various features of Bevy UI
```

It won't work the way you want to otherwise ;)

**Comment by cart** [comment]
I would review it, but given that I haven't reviewed the current batching pr yet, maybe hold off until after. There are a few corner cases that I'm worried about that I haven't tested on the current impl.

**Comment by Davier** [comment]
Text and Text2d are now implemented. The ui_pipelined and text2d_pipelined examples seem to work correctly.
Should we start to review and merge now so that more people can try out the new renderer? Or wait for the other in-flight PRs to be merged (shaders as assets, sprite batching, MSAA, ...)?

**Comment by Davier** [comment]
Updated on top of pipelined-rendering. The first commit is a simple copy of the old bevy_ui and bevy_text to make reviewing changes easier.
I used the last changes in bevy_sprites, including batching.
I'll rebase on top of #3175 when it's merged.

Remaining issues/questions:
1. the background is not cleared with ClearColor if there is no 2d or 2s camera, in that case the UI is drawn on top of garbage.
2. the VisibleEntities component is required for a camera to work, even if the UI camera has no culling
3. should the UiPipeline specialization be removed since the key is empty? I kept it to make it easier to add new specializations in the future.
4. should the UI renderer use MSAA?

Possible future improvements:
1. clipping
2. texture atlases
3. opaque / transparent / alpha mask phases
4. use depth buffer to avoid overdraw
5. include flags in the vertex buffer and skip sampling the texture when possible (like for PBR)

**Comment by cart** [comment]
Changed the base branch to `main` because thats where development is happening now!

**Review by colepoirier — APPROVED** [review]
I approve, but someone with more bevy experience and/or more experience with the new renderer will need to do the final sign off. I don’t see any obvious problems but I don’t think I would be able to see a great deal of the problems if they do exist.

File: pipelined/bevy_ui2/src/ui_node.rs
Comment (Davier) [inline pipelined/bevy_ui2/src/ui_node.rs:279]: The names of these new components are horrible since they alias the types they contain. Any suggestions?

File: pipelined/bevy_ui2/src/update.rs
Comment (Davier) [inline pipelined/bevy_ui2/src/update.rs:54]: I adapted the test to `derive(Component)` independently but it was already done in main, I should replace this file with `crates/bevy_ui/src/update.rs`.

**Review by Davier — COMMENTED** [review]
I did another full self review. I caught a few things and detailed a few others, hopefully it will help other reviewers.
I recommend selecting all but the first commit in the filters, to view what actually changed.

File: pipelined/bevy_ui2/src/entity.rs
Comment (bjorn3) [reply pipelined/bevy_ui2/src/entity.rs:97]: In case of a scrollable element part of the UI items are likely outside of the bounds of the scrollable element and thus must be invisible.

Comment (Davier) [reply pipelined/bevy_ui2/src/entity.rs:97]: I'm tackling that case in another (soon to be) PR for clipping, where I implement the `Style::overflow` property ([it is currently commented out](https://github.com/bevyengine/bevy/blob/f4cfcc0e44b91446beb49a1dbe9965cb7bcde059/crates/bevy_ui/src/ui_node.rs#L229-L241)). I calculate how items are clipped in the queue_uinode fn, and skip those completely clipped. It instead might be possible to calculate it beforehand and setting a `ComputedVisibility` accordingly.

**Comment by Davier** [comment]
There is an issue with updating images (including text glyphs), like for sprites (see #3207). Whichever solution is chosen for sprite should also be applied for UI.

File: pipelined/bevy_ui2/src/ui_node.rs
Comment (TheRawMeatball) [reply pipelined/bevy_ui2/src/ui_node.rs:279]: Prepend `Ui` to the names? So like `UiColor`, `UiImage` etc.

Comment (Davier) [reply pipelined/bevy_ui2/src/ui_node.rs]: Done in c8975e5

File: pipelined/bevy_ui2/src/update.rs
Comment (Davier) [reply pipelined/bevy_ui2/src/update.rs:54]: Done in faacdd1

**Comment by Davier** [comment]
> There is an issue with updating images (including text glyphs), like for sprites (see #3207). Whichever solution is chosen for sprite should also be applied for UI.

Fixed in a6f1581 (after rebasing on top of the fix for sprites). I chose to reuse `SpriteAssetEvents`, since the bevy_ui2 already depends on bevy_sprite2. Maybe that resource should be renamed to be more generic, or even placed in bevy_render2 or bevy_core_pipeline?

File: pipelined/bevy_ui2/src/ui_node.rs
Comment (TheRawMeatball) [inline pipelined/bevy_ui2/src/ui_node.rs:282]: I'd like to suggest we use just an `Handle<Image>` without an option, and use a 1x1 image as the default for the handle. Alternatively, we could make the handle a weak handle that points to nothing.

**Comment by TheRawMeatball** [comment]
The last commit could likely be reverted, as in retrospect, I don't think the ControlNode s are a great idea, and were merged too hastily.

File: pipelined/bevy_ui2/src/ui_node.rs
Comment (Davier) [reply pipelined/bevy_ui2/src/ui_node.rs:274]: A default handle for images was just added in #3270, I'll convert this PR to use it too.

**Comment by cart** [comment]
I'm diving in to this now. Tip for reviewing: deleting the old `bevy_ui` and `bevy_text` and replacing them with the pipelined crates makes it much easier to review these changes because you can see the actual diffs.

**Comment by mockersf** [comment]
or in GitHub UI, there's is a dropdown when viewing file changes saying "changes from all commits" where you can [select every commit but the first one](https://github.com/bevyengine/bevy/pull/2653/files/ee2ccaf4a82767656befcc3719055e78363c8a1c..911952fa153d9d1be2459a890813c287de2f520e) (the url won't be valid if more commits are pushed)

File: pipelined/bevy_ui2/Cargo.toml
Comment (cart) [inline pipelined/bevy_ui2/Cargo.toml:35]: New render features generally shouldn't need to add a direct wgpu dependency. If we're missing wgpu exports, can you add them to the existing re-exports in bevy_render2::render_resources?

File: pipelined/bevy_ui2/src/entity.rs
Comment (cart) [inline pipelined/bevy_ui2/src/entity.rs:17]: Part of me wants to just merge UiColor and UiImage into `Node`. I really don't like wrapper structs, and we're adding these to all Nodes anyway. I think that would improve UX.

**Review by cart — COMMENTED** [review]
I like this port! Generally 1:1 with a number of nice tweaks here and there. I also like that you borrowed the batched sprite rendering impl!

File: pipelined/bevy_ui2/src/entity.rs
Comment (cart) [reply pipelined/bevy_ui2/src/entity.rs:17]: Although I guess Node could be used in a lot of non-color + image contexts. I guess the `color.into()` pattern isn't terrible. We could also merge UiColor and UiImage into a single component and then used name fields, but that does make this less flexible (some widgets might want a color, but not an image). I guess lets leave this as is for now and we can revisit later if we can think of something better.

File: pipelined/bevy_ui2/src/ui_node.rs
Comment (cart) [reply pipelined/bevy_ui2/src/ui_node.rs:274]: Yeah I like this

File: pipelined/bevy_ui2/src/render/render_pass.rs
Comment (cart) [reply pipelined/bevy_ui2/src/render/render_pass.rs:151]: The bind group is used globally (because we use a dynamic uniform). No need to store it on each individual view. We already store the offset as a per-view component, which feels right to me.

File: pipelined/bevy_ui2/src/render/mod.rs
Comment (cart) [reply pipelined/bevy_ui2/src/render/mod.rs:346]: I think for now its fine to embrace the decoupling here. Can we rename this to UiImageBindGroups so it doesn't get conflated with `bevy_sprite2::ImageBindGroups`?

File: examples/ui/ui_pipelined.rs
Comment (cart) [reply examples/ui/ui_pipelined.rs:28]: Merged!

**Comment by cart** [comment]
Just merged #3209. Lets resolve conflicts, make sure this plays nicely with the new clearing behavior, and address the remaining open comments. Then i think this is ready to go!

File: pipelined/bevy_ui2/src/ui_node.rs
Comment (Davier) [reply pipelined/bevy_ui2/src/ui_node.rs]: Done in 67ae82e

File: pipelined/bevy_ui2/src/render/mod.rs
Comment (Davier) [reply pipelined/bevy_ui2/src/render/mod.rs]: Done in 41ec574

File: pipelined/bevy_ui2/Cargo.toml
Comment (Davier) [reply pipelined/bevy_ui2/Cargo.toml]: Done in 8c19df4, it seems everything was exported correctly (some might have been missing early on).

File: examples/ui/ui_pipelined.rs
Comment (Davier) [reply examples/ui/ui_pipelined.rs]: Removed in e8ab002, it seems to work correctly.

**Comment by Davier** [comment]
Unless there is more feedback on `Visibility`, everything is addressed. I only lightly tested the new modifications. I'll play a bit more with it tomorrow, but it already looks good to merge if you want to.

**Comment by cart** [comment]
Changes look good to me. I'm gonna think a bit on the visibility situation for a bit, but I think this is probably good to merge!

**Comment by cart** [comment]
I'm fine with the choices you made regarding visibility. An empty `VisibleEntities` on the ui camera doesnt bother me and it leaves the doors open to future UI visibility work. The fact that UI elements dont have visibility / computed visibilty atm is also ok by me (for now).

**Pull request closed without merging**
